Adds corrections for Julia v1.9

apalha opened the pull request:
I tried to use MeshCore on Julia v1.9.2 and I ran into some internal errors in MeshCore. After some debugging, these changes solved the problem:

- Adds TupleTools to enable sorting of tuple, direct sorting over a tuple was not working in v1.9.
- Adds explicit elementwise operations that were not working in v1.9.
- Adds explicit access to StaticArray data to enable some operations, also not working in v1.9.

Would be nice to have this available in the official release.

Thanks!

-artur palha

apalha committed:
Adds corrections for Julia v1.9
- Adds TupleTools to enable sorting of tuple, direct sorting over a tuple was not working in v1.9.
- Adds explicit elementwise operations that were not working in v1.9.
- Adds explicit access to StaticArray data to enable some operations, also not working in v1.9.

PetrKryslUCSD commented:
Could you be more specific what the errors were please? 
```
julia> versioninfo()                                                                                                                                                                         
Julia Version 1.9.2                                                                                                                                                                          
Commit e4ee485e90 (2023-07-05 09:39 UTC)                                                                                                                                                     
Platform Info:                                                                                                                                                                               
  OS: Windows (x86_64-w64-mingw32)                                                                                                                                                           
  CPU: 8 × Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz                                                                                                                                         
  WORD_SIZE: 64                                                                                                                                                                              
  LIBM: libopenlibm                                                                                                                                                                          
  LLVM: libLLVM-14.0.6 (ORCJIT, icelake-client)                                                                                                                                              
  Threads: 4 on 8 virtual cores                                                                                                                                                              
Environment:                                                                                                                                                                                 
  JULIA_IMAGE_THREADS = 1                                                                                                                                                                    
  JULIA_DEPOT_PATH = C:/Users/pkonl/SublimeJulia-1.9.2/assets/.julia-1.9.2-depot                                                                                                             
  JULIA_SSL_CA_ROOTS_PATH =                                                                                                                                                                  
  JULIA_EDITOR = C:/Users/pkonl/SublimeJulia-1.9.2/assets/st4/subl                                                                                                                           
                                                                                                                                                                          
julia> using Pkg; Pkg.status();                                                                                                                                                              
Project MeshCore v1.3.2                                                                                                                                                                      
Status `C:\Users\pkonl\Documents\00WIP\MeshCore.jl\Project.toml`                                                                                                                             
  [98e50ef6] JuliaFormatter v1.0.34                                                                                                                                                          
  [a5566ce3] REPLHistory v0.2.0                                                                                                                                                              
  [295af30f] Revise v3.5.3                                                                                                                                                                   
  [90137ffa] StaticArrays v1.6.2                                                                                                                                                             
  [8dfed614] Test                                                                                                                                                                            
                                                                                                                                                                                             
```
So, tests pass. Please advise what it is you are seeing. Thanks.

apalha commented:
Yes, of course. I will reverse the changes and post the errors here.

apalha commented:
This was the first error I got, which led to the first source change. Once I changed "=" by ".=" I got another error, the sorting of a tuple, which I corrected by adding TupleTools, and then accessing the data of the arrays by .data.

It is strange that you do not have the same errors. I am on macOS with an arm processor, but this should not make any difference.

`ERROR: LoadError: ArgumentError: indexed assignment with a single value to possibly many locations is not supported; perhaps use broadcasting `.=` instead?
Stacktrace:
 [1] setindex_shape_check(::NTuple{4, Int64}, ::Int64, ::Int64)
   @ Base ./indices.jl:261
 [2] _unsafe_setindex!(::IndexLinear, ::Matrix{Int64}, ::NTuple{4, Int64}, ::Int64, ::Base.Slice{Base.OneTo{Int64}})
   @ Base ./multidimensional.jl:925
 [3] _setindex!
   @ ./multidimensional.jl:916 [inlined]
 [4] setindex!
   @ ./abstractarray.jl:1397 [inlined]
 [5] _asmatrix
   @ ~/.julia/packages/MeshCore/K4gV7/src/increl.jl:191 [inlined]
 [6] ir_skeleton(ir::MeshCore.IncRel{MeshCore.Q4ShapeDesc{2, 4, 4, MeshCore.L2ShapeDesc, 4, MeshCore.P1ShapeDesc, 4, 4}, MeshCore.P1ShapeDesc{0, 1, 0, MeshCore.NoSuchShapeDesc, 0, MeshCore.NoSuchShapeDesc, 1, 0}, NTuple{4, Int64}}, name::String)
   @ MeshCore ~/.julia/packages/MeshCore/K4gV7/src/increl.jl:268
 [7] ir_skeleton(ir::MeshCore.IncRel{MeshCore.Q4ShapeDesc{2, 4, 4, MeshCore.L2ShapeDesc, 4, MeshCore.P1ShapeDesc, 4, 4}, MeshCore.P1ShapeDesc{0, 1, 0, MeshCore.NoSuchShapeDesc, 0, MeshCore.NoSuchShapeDesc, 1, 0}, NTuple{4, Int64}})
   @ MeshCore ~/.julia/packages/MeshCore/K4gV7/src/increl.jl:265
 [8] top-level scope
   @ ~/Dropbox/Documents/Work/courses/deep_intro_to_julia/meshcore.jl:9
in expression starting at /Users/apalha/Dropbox/Documents/Work/courses/deep_intro_to_julia/meshcore.jl:9`

apalha commented:
Maybe also relevant is the code that triggers this error, not sure if your tests cover the whole source:

```julia
import MeshCore

q4s = MeshCore.ShapeColl(MeshCore.Q4, 6)
vrts = MeshCore.ShapeColl(MeshCore.P1, 12)

c = [(1, 2, 6, 5), (5, 6, 10, 9), (6, 2, 3, 7), (6, 7, 11, 10), (3, 4, 8, 7), (7, 8, 12, 11)]
d2_d0 = MeshCore.IncRel(q4s, vrts, c)

d1_d0 = MeshCore.ir_skeleton(d2_d0)
```

The last line raises the error.

PetrKryslUCSD commented:
I see what is going on: The incidence relations should be built from a vector of vectors, not from a vector of tuples. 

But, it would be a good extension to allow this, and I will consider this PR.

What was not working with StaticArrays?

Thank you.

PetrKryslUCSD commented:
I see what is going on: The incidence relations should be built from a vector of vectors, not from a vector of tuples. 

But, it would be a good extension to allow this, and I will consider this PR.

What was not working with StaticArrays?

apalha commented:
I will place here one by one the errors I got and the corresponding source code changes I propose after the .= change above.

1.

```julia
import MeshCore

q4s = MeshCore.ShapeColl(MeshCore.Q4, 6)
vrts = MeshCore.ShapeColl(MeshCore.P1, 12)

c = [(1, 2, 6, 5), (5, 6, 10, 9), (6, 2, 3, 7), (6, 7, 11, 10), (3, 4, 8, 7), (7, 8, 12, 11)]

d2_d0 = MeshCore.IncRel(q4s, vrts, c)
d1_d0 = MeshCore.ir_skeleton(d2_d0)
d0_d1 = MeshCore.ir_transpose(d1_d0)
d2_d1 = MeshCore.ir_bbyfacets(d2_d0, d1_d0, d0_d1)  # after the .= correction the next error takes place here
```

This is the error:

```julia
ERROR: LoadError: MethodError: no method matching sort(::Tuple{Int64, Int64})

Closest candidates are:
  sort(::AbstractUnitRange)
   @ Base range.jl:1384
  sort(::AbstractRange)
   @ Base range.jl:1387
  sort(::OrderedCollections.LittleDict; byvalue, args...)
   @ OrderedCollections ~/.julia/packages/OrderedCollections/SInLM/src/dict_sorting.jl:38
  ...

Stacktrace:
 [1] ir_bbyfacets(ir::MeshCore.IncRel{MeshCore.Q4ShapeDesc{2, 4, 4, MeshCore.L2ShapeDesc, 4, MeshCore.P1ShapeDesc, 4, 4}, MeshCore.P1ShapeDesc{0, 1, 0, MeshCore.NoSuchShapeDesc, 0, MeshCore.NoSuchShapeDesc, 1, 0}, NTuple{4, Int64}}, fir::MeshCore.IncRel{MeshCore.L2ShapeDesc{1, 2, 2, MeshCore.P1ShapeDesc, 0, MeshCore.NoSuchShapeDesc, 2, 0}, MeshCore.P1ShapeDesc{0, 1, 0, MeshCore.NoSuchShapeDesc, 0, MeshCore.NoSuchShapeDesc, 1, 0}, StaticArraysCore.SVector{2, Int64}}, tfir::MeshCore.IncRel{MeshCore.P1ShapeDesc{0, 1, 0, MeshCore.NoSuchShapeDesc, 0, MeshCore.NoSuchShapeDesc, 1, 0}, MeshCore.L2ShapeDesc{1, 2, 2, MeshCore.P1ShapeDesc, 0, MeshCore.NoSuchShapeDesc, 2, 0}, Vector{Int64}}, name::String)
   @ MeshCore ~/.julia/packages/MeshCore/K4gV7/src/increl.jl:376
 [2] ir_bbyfacets(ir::MeshCore.IncRel{MeshCore.Q4ShapeDesc{2, 4, 4, MeshCore.L2ShapeDesc, 4, MeshCore.P1ShapeDesc, 4, 4}, MeshCore.P1ShapeDesc{0, 1, 0, MeshCore.NoSuchShapeDesc, 0, MeshCore.NoSuchShapeDesc, 1, 0}, NTuple{4, Int64}}, fir::MeshCore.IncRel{MeshCore.L2ShapeDesc{1, 2, 2, MeshCore.P1ShapeDesc, 0, MeshCore.NoSuchShapeDesc, 2, 0}, MeshCore.P1ShapeDesc{0, 1, 0, MeshCore.NoSuchShapeDesc, 0, MeshCore.NoSuchShapeDesc, 1, 0}, StaticArraysCore.SVector{2, Int64}}, tfir::MeshCore.IncRel{MeshCore.P1ShapeDesc{0, 1, 0, MeshCore.NoSuchShapeDesc, 0, MeshCore.NoSuchShapeDesc, 1, 0}, MeshCore.L2ShapeDesc{1, 2, 2, MeshCore.P1ShapeDesc, 0, MeshCore.NoSuchShapeDesc, 2, 0}, Vector{Int64}})
   @ MeshCore ~/.julia/packages/MeshCore/K4gV7/src/increl.jl:352
 [3] top-level scope
   @ ~/Dropbox/Documents/Work/courses/deep_intro_to_julia/meshcore.jl:17
in expression starting at /Users/apalha/Dropbox/Documents/Work/courses/deep_intro_to_julia/meshcore.jl:17
```
and the offending line was the line (376):

```julia
sfc = sort(fc)
```

inside `ir_bbyfacets(ir::IncRel, fir::IncRel, tfir::IncRel, name = "bbf")`

Essentially it is not able to sort a tuple. To sort a tuple we can use TupleTools, so to solve this issue I
added TupleTools to the dependences and imported the sort function:

Project.toml

```julia
[deps]
TupleTools = "9d95972d-f1c8-5527-a6e0-b4b365fa01f6"  # line added
StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
```

src/MeshCore.jl

```julia
using TupleTools: sort  # the sort for tuples is required because not sorting for tuples is available
```

2. Once TupleTools is added, then the following error appears

```julia
ERROR: LoadError: MethodError: no method matching sort(::StaticArraysCore.SVector{2, Int64})
You may have intended to import Base.sort

Closest candidates are:
  sort(::Tuple; lt, by, rev)
   @ TupleTools ~/.julia/packages/TupleTools/ecrzx/src/TupleTools.jl:243

Stacktrace:
 [1] ir_bbyfacets(ir::MeshCore.IncRel{MeshCore.Q4ShapeDesc{2, 4, 4, MeshCore.L2ShapeDesc, 4, MeshCore.P1ShapeDesc, 4, 4}, MeshCore.P1ShapeDesc{0, 1, 0, MeshCore.NoSuchShapeDesc, 0, MeshCore.NoSuchShapeDesc, 1, 0}, NTuple{4, Int64}}, fir::MeshCore.IncRel{MeshCore.L2ShapeDesc{1, 2, 2, MeshCore.P1ShapeDesc, 0, MeshCore.NoSuchShapeDesc, 2, 0}, MeshCore.P1ShapeDesc{0, 1, 0, MeshCore.NoSuchShapeDesc, 0, MeshCore.NoSuchShapeDesc, 1, 0}, StaticArraysCore.SVector{2, Int64}}, tfir::MeshCore.IncRel{MeshCore.P1ShapeDesc{0, 1, 0, MeshCore.NoSuchShapeDesc, 0, MeshCore.NoSuchShapeDesc, 1, 0}, MeshCore.L2ShapeDesc{1, 2, 2, MeshCore.P1ShapeDesc, 0, MeshCore.NoSuchShapeDesc, 2, 0}, Vector{Int64}}, name::String)
   @ MeshCore ~/.julia/dev/MeshCore/K4gV7/src/increl.jl:379
 [2] ir_bbyfacets(ir::MeshCore.IncRel{MeshCore.Q4ShapeDesc{2, 4, 4, MeshCore.L2ShapeDesc, 4, MeshCore.P1ShapeDesc, 4, 4}, MeshCore.P1ShapeDesc{0, 1, 0, MeshCore.NoSuchShapeDesc, 0, MeshCore.NoSuchShapeDesc, 1, 0}, NTuple{4, Int64}}, fir::MeshCore.IncRel{MeshCore.L2ShapeDesc{1, 2, 2, MeshCore.P1ShapeDesc, 0, MeshCore.NoSuchShapeDesc, 2, 0}, MeshCore.P1ShapeDesc{0, 1, 0, MeshCore.NoSuchShapeDesc, 0, MeshCore.NoSuchShapeDesc, 1, 0}, StaticArraysCore.SVector{2, Int64}}, tfir::MeshCore.IncRel{MeshCore.P1ShapeDesc{0, 1, 0, MeshCore.NoSuchShapeDesc, 0, MeshCore.NoSuchShapeDesc, 1, 0}, MeshCore.L2ShapeDesc{1, 2, 2, MeshCore.P1ShapeDesc, 0, MeshCore.NoSuchShapeDesc, 2, 0}, Vector{Int64}})
   @ MeshCore ~/.julia/dev/MeshCore/K4gV7/src/increl.jl:352
 [3] top-level scope
   @ ~/Dropbox/Documents/Work/courses/deep_intro_to_julia/meshcore.jl:17
in expression starting at /Users/apalha/Dropbox/Documents/Work/courses/deep_intro_to_julia/meshcore.jl:17
```

The offending lines (379 and 380) are: 

```julia
if sfc == sort(oc)
    sgn = _sense(fc[1:n1st], oc, nshif)
```
Which has to do with trying to sort a static array, which is not available. If we make the sort on `oc.data` then it works (that is the correction I propose). Also, after that, in the line after the sort, we also need to change `oc` by `oc.data` for similar reasons.

PetrKryslUCSD commented:
Please try the latest registered version, 1.3.3. It runs your examples.

apalha commented:
No errors now, results seems good.

Thanks for quickly looking into this and fixing it.

Kind regards,

-artur palha

PetrKryslUCSD commented:
No problem. Do let me know if anything else is needed. A star will be appreciated. :-)
